docs/glossary.rst

```diff
@@ -0,0 +1,274 @@
+          their degree of completion and, potentially, with references to other
+          resources to describe intermediate or final results.
+
+    port
+        Generic term for a named source, sink, resource, or binding hook on a node.
+
+    resource
+        Describes an API hook for an interaction mediated by a Context. Data flow
+        is described as *immutable* resources (generally produced as Operation outputs)
+        that can be consumed by binding to Operation inputs or by extracting as *results*
+        from the API. Some interactions cannot be represented in terms of producers
+        and subscribers of immutable data events: *Mutable* resources cannot be
+        managed by the Context as data events and require different work scheduling
+        policies that either (a) allows arbitrary (unscheduled) call-back through the API framework,
+        (b) dispatch the mutable resource collaboration to another Context, or (c)
+        allow operations to bind and interact with an interface not specified by the
+        API or not known to the responsible Context implementation. Examples include
+        the Context-provided *ensemble_reduce* functionality, the ensemble simulation
+        signaling facility (by which extension code can terminate a simulation early),
+        and the binding mechanism by which MD extension code can be attached to an
+        *MD* operation as a plugin. The nature of a resource is indicated by the
+        namespace of its *port* in the work record.
+
+    run time
+        .. todo:: Define the scope to be conveyed by the noun "run time".
+
+    simulation segment
+    trajectory segment
+        A sequence of molecular simulation iterations or frames produced
+        deterministically (within numerical limits) under well-determined
+        parameters. For the purposes of discussing checkpoint intervals or the
+        minimum amount of work executed between API calls, it is useful to
+        distinguish between full simulation trajectories and the irreducible
+        unit of work supported by a simulation library. In the simplest API use
+        cases, a simulation library does not interact with the API during
+        production of a simulation segment, and allows for reinitialization
+        between simulation segments. This allows for unambiguous labeling of the
+        artifacts of a segment. Optimizations may focus on reducing overhead
+        between successive simulation segments (minimizing reinitialization).
+        Extensions may introduce abstractions for well-characterized non-constant
+        parameters, such as time-varying lambda values, though such abstractions
+        are not required in the API since the effect can be achieved through
+        binding to a mutable resource (with details beyond the scope of the API)
+        owned by another operation whose state and action is well characterized
+        for the segment.
+
+    task
```

Comment:
@mrshirts writes:
> So with Markov State Modeling, the task would be "Perform MD simulations for X steps of this given system with these MD parameters?" The use of "operation" within the task definition is confusing here.

My intention is that the _operation_ is the representation of "perform this (batch of) MD simulation(s) to generate a specific (set of) well-characterized trajectory segment(s) with this parameter/input set."

The _task_ or _tasks_ would be the one or more function calls the worker is told to make to carry out the operation. I'm attempting to build a connection between the high level representation of work and the run time scheduling/execution details.

Reply:
I want to defer the definition of "task" to @mturilli and @andre-merzky so that we can answer these questions. I think we should have a goal of defining the relationship between "task" and "operation" or "simulation segment," but I don't think we have reached a point where we can do that with finality.

Reply:
Task: a unit of work

Reply:
@andre-merzky is a "task" necessarily atomic or is it potentially decomposable and/or recoverably stateful? Is it a unique representation of equivalent work? Or might a high-level representation of work be translated into different task (or tasks) under different circumstances, such as for different execution environments?

My expectation is that a "task" is the unit of work that is relevant to the run time, whereas an "operation" is the unit of work that is relevant to the highest-level API, and that the API defers the mapping of "operations" to "tasks" to lower layers. Somewhere in the middle, a simulation operation with intermediate structure (checkpoints, call-back intervals, etc.) would need to be considered in terms of a series of simulation segments, but the mapping to "tasks" depends on how it is executed, and the context in which we want to talk about "tasks." (I.e. does a "task" map specifically to one CPI or RPC call? Is it a function object? A scheduling unit?)

But I think the important thing here is to note that "task" is a unit of executable work in the domain of the execution manager and/or run time. So how about

"task: The unit of work when dispatching operations for execution."

Reply:
I tend not to define terms so tightly.  I am not trying to say that terms should not be well defined - but when pinning `task` to `The unit of work when dispatching operations for execution`, then that would mean I need a different term when talking about `The unit of work when *preparing* operations for execution`.  I'd instead use qualifiers: atomic task, checkpointable task, etc.

Not sure if that helps.  As said earlier (by Matteo), I don't think we necessarily need to reconcile our glossaries right now, so I really don't want to make a proposal here.  Mostly just replying because you prompted directly :-)